protocols/imap/src/main/java/org/apache/james/imap/processor/StoreProcessor.java

```diff
@@ -264,50 +265,51 @@ private void setFlags(StoreRequest request, MailboxSession mailboxSession, Messa
-                if (silent && (unchangedSince != -1 || qresyncEnabled || condstoreEnabled)) {
-                    // We need to return an FETCH response which contains the mod-sequence of the message even if FLAGS.SILENT was used
-                    response = new FetchResponse(msn, null, resultUid, modSeqs.get(uid), null, null, null, null, null, null);
-                } else if (!silent && (unchangedSince != -1 || qresyncEnabled || condstoreEnabled)) {
+                    // Check if we need to include the uid. T
                     //
-                    // Use a FETCH response which contains the mod-sequence and the flags
-                    response = new FetchResponse(msn, resultFlags, resultUid, modSeqs.get(uid), null, null, null, null, null, null);
-                } else {
-                    // Use a FETCH response which only contains the flags as no CONDSTORE was used
-                    response = new FetchResponse(msn, resultFlags, resultUid, null, null, null, null, null, null, null);
-                }
-                responder.respond(response);
+                    // This is the case if one of these is true:
+                    //      - FETCH (UID...)  was used
+                    //      - QRESYNC was enabled via ENABLE QRESYNC
+                    if (useUids || qresyncEnabled) {
+                        resultUid = uid;
+                    } else {
+                        resultUid = null;
+                    }
+
+                    if (selected.isRecent(uid)) {
+                        resultFlags.add(Flags.Flag.RECENT);
+                    }
+
+                    final FetchResponse response;
+                    // For more information related to the FETCH response see
+                    //
+                    // RFC4551 3.2. STORE and UID STORE Commands
+                    if (silent && (unchangedSince != -1 || qresyncEnabled || condstoreEnabled)) {
+                        // We need to return an FETCH response which contains the mod-sequence of the message even if FLAGS.SILENT was used
+                        response = new FetchResponse(msn, null, resultUid, modSeqs.get(uid), null, null, null, null, null, null);
+                    } else if (!silent && (unchangedSince != -1 || qresyncEnabled || condstoreEnabled)) {
+                        //
+                        // Use a FETCH response which contains the mod-sequence and the flags
+                        response = new FetchResponse(msn, resultFlags, resultUid, modSeqs.get(uid), null, null, null, null, null, null);
+                    } else {
+                        // Use a FETCH response which only contains the flags as no CONDSTORE was used
+                        response = new FetchResponse(msn, resultFlags, resultUid, null, null, null, null, null, null, null);
+                    }
+                    responder.respond(response);
+                    return null;
+                });
```

Comment:
should be more readable if you perform an extraction in a commit before

Reply:
well, it's just a given problem, if you enable whitespace changes filtering, everything becomes readable

Reply:
right, in this case it helps really (in an other PR I had more issues than it solves)